File: Sources/JVoice/Services/Orchestration/TranscriptHistoryStore.swift
import Foundation

/// One persisted transcript in the recent-history list. `id` gives SwiftUI a
/// stable identity for rows (delete-by-id) independent of the text.
struct TranscriptEntry: Codable, Identifiable, Equatable {
    let id: UUID
    let text: String

    init(id: UUID = UUID(), text: String) {
        self.id = id
        self.text = text
    }
}

/// Remembers the most recent transcripts (newest first), capped at `maxEntries`.
///
/// Persisted as JSON in `UserDefaults` (same plaintext-in-prefs posture as
/// `LastTranscriptStore`): a few dozen short dictation snippets is only a few KB,
/// disk-backed and lazily loaded, so nothing heavy stays resident. Privacy: the
/// list is only ever erased by explicit user action (the "Clear all"/per-row
/// delete controls, or "Restore Default Settings").
final class TranscriptHistoryStore {
    static let maxEntries = 30

    private let defaults: UserDefaults
    private static let key = "jvoice.app.transcriptHistory"

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
    }

    /// Newest first. A corrupt/absent blob decodes to an empty list (same as a
    /// fresh install) rather than throwing.
    var entries: [TranscriptEntry] {
        guard let data = defaults.data(forKey: Self.key),
              let decoded = try? JSONDecoder().decode([TranscriptEntry].self, from: data)
        else { return [] }
        return decoded
    }

    /// Prepend a new transcript and trim to `maxEntries`, dropping the oldest.
    /// Blank text is ignored.
    @discardableResult
    func add(_ text: String) -> [TranscriptEntry] {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return entries }
        var list = entries
        list.insert(TranscriptEntry(text: trimmed), at: 0)
        if list.count > Self.maxEntries {
            list.removeLast(list.count - Self.maxEntries)
        }
        persist(list)
        return list
    }

    @discardableResult
    func remove(id: UUID) -> [TranscriptEntry] {
        var list = entries
        list.removeAll { $0.id == id }
        persist(list)
        return list
    }

    func clear() {
        defaults.removeObject(forKey: Self.key)
    }

    private func persist(_ list: [TranscriptEntry]) {
        if let data = try? JSONEncoder().encode(list) {
            defaults.set(data, forKey: Self.key)
        }
    }
}

File: windows/JVoice.App/Platform/Persistence/TranscriptHistoryStore.cs
using System.IO;
using JVoice.Core.Models;

namespace JVoice.App.Platform;

/// Persists the Recent Transcripts history (the most recent finalized dictations,
/// newest first) to %APPDATA%\JVoice\transcript-history.json.
///
/// Windows-only feature. Mirrors the StatsStore / LastTranscriptStore shape: the pure
/// list + JSON logic lives in JVoice.Core (TranscriptHistory, unit-tested); this wrapper
/// only does file I/O. Persists synchronously on each mutation (a few dozen short snippets
/// is only a few KB, written rarely — once per dictation or per edit).
///
/// Privacy: stored in plaintext, and the list is erased ONLY by explicit user action
/// (per-row delete, Clear all, or Restore Default Settings) — never automatically. A
/// missing or corrupt file loads as an empty list, exactly like a fresh install.
public sealed class TranscriptHistoryStore
{
    private readonly string _path;
    private readonly object _gate = new();
    private List<TranscriptHistoryEntry> _entries;

    public TranscriptHistoryStore(string? path = null)
    {
        _path = path ?? PlatformPaths.TranscriptHistoryFile;
        _entries = TranscriptHistory.Deserialize(ReadAllOrNull(_path)).ToList();
    }

    /// Current entries, newest first. A snapshot copy — callers can't mutate our list.
    public IReadOnlyList<TranscriptHistoryEntry> Entries
    {
        get { lock (_gate) return _entries.ToList(); }
    }

    /// Prepend a finalized transcript. Trims whitespace; blank text is ignored and
    /// returns null. Caps the list at TranscriptHistory.MaxEntries and persists.
    /// Returns the newly-added entry so the UI can insert just that row.
    public TranscriptHistoryEntry? Add(string text)
    {
        lock (_gate)
        {
            var (list, added) = TranscriptHistory.Add(_entries, text, Guid.NewGuid());
            if (added is null) return null;
            _entries = list.ToList();
            Save();
            return added;
        }
    }

    /// Remove the one entry with this id (if present) and persist.
    public void Remove(Guid id)
    {
        lock (_gate)
        {
            var updated = TranscriptHistory.Remove(_entries, id);
            if (updated.Count == _entries.Count) return; // nothing matched
            _entries = updated.ToList();
            Save();
        }
    }

    /// Empty the whole history and persist.
    public void Clear()
    {
        lock (_gate)
        {
            if (_entries.Count == 0) return;
            _entries = [];
            Save();
        }
    }

    private static string? ReadAllOrNull(string path)
    {
        try { return File.Exists(path) ? File.ReadAllText(path) : null; }
        catch { return null; }
    }

    private void Save()
    {
        try
        {
            Directory.CreateDirectory(Path.GetDirectoryName(_path)!);
            string json = TranscriptHistory.Serialize(_entries);
            // Atomic-ish write: temp file then move, so a crash mid-write can't truncate
            // the history file (same pattern as SettingsStore / StatsStore).
            string tmp = _path + ".tmp";
            File.WriteAllText(tmp, json);
            File.Move(tmp, _path, overwrite: true);
        }
        catch (Exception ex)
        {
            SystemActions.ReportError($"Failed to save transcript history. {ex.Message}");
        }
    }
}
